### PR TITLE
Refactored Desktop class to support multiple desktop instances at the same time

### PR DESCRIPTION
**Pull request by maxcapodi78**
Since this functionality is not supported officially by AEDT GRPC, there are some limitations in the implementation:
- Release desktop only is not applicable when multiple desktop are connected at the same time
- Desktop have to be closed in reverse order to creation (LIFO)

### CODE REVIEW

File: pyaedt/desktop.py [hunk 1]
```diff
@@ -267,23 +239,47 @@ def _close_aedt_application(close_desktop, pid, is_grpc_api):
...
+                if _desktop_sessions:
+                    for v in _desktop_sessions.values():
+                        if pid in v.parent_desktop_id:
+                            del v.parent_desktop_id[v.parent_desktop_id.index(pid)]
+                return True
             except:  # pragma: no cover
                 warnings.warn("Something went wrong closing AEDT. Exception in `_main.oDesktop.QuitApplication()`.")
                 pass
+        elif _desktop_sessions and len(_desktop_sessions) > 1 and not desktop_class.parent_desktop_id:
+            pyaedt_logger.error("Release is not allowed when multiple Desktop sessions are available.")
```
Comment (PipKat) [inline pyaedt/desktop.py:260]:
```suggestion
            pyaedt_logger.error("Release is not allowed when multiple desktop sessions are available.")
```

File: pyaedt/desktop.py [hunk 2]
```diff
@@ -267,23 +239,47 @@ def _close_aedt_application(close_desktop, pid, is_grpc_api):
...
+                    for v in _desktop_sessions.values():
+                        if pid in v.parent_desktop_id:
+                            del v.parent_desktop_id[v.parent_desktop_id.index(pid)]
+                return True
             except:  # pragma: no cover
+                warnings.warn("Something went wrong closing AEDT. Exception in `_main.oDesktop.QuitApplication()`.")
                 pass
+        elif _desktop_sessions and len(_desktop_sessions) > 1:
+            pyaedt_logger.error("A child desktop session is linked to this session.")
+            pyaedt_logger.error("Multiple Desktop sessions have to be released in reverse order.")
```
Comment (PipKat) [inline pyaedt/desktop.py:274]:
```suggestion
            pyaedt_logger.error("Multiple desktop sessions must be released in reverse order.")
```

File: pyaedt/generic/design_types.py [hunk 3]
```diff
@@ -1750,45 +1750,63 @@ def Siwave(
...
     Parameters
     ----------
-    project_name
-    design_name
+    project_name : str, optional
+        Project name.
+    design_name : str, optional
+        Design name.
+    desktop :  :class:`pyaedt.desktop.Desktop`, optional
+        Desktop class.
```
Comment (PipKat) [inline pyaedt/generic/design_types.py:1763]:
```suggestion
    desktop : :class:`pyaedt.desktop.Desktop`, optional
        Desktop class. The default is ``None``.
```
What behavior occurs when the default of ``None`` is used. Should we indicate it in this format: "The default is ``None``, in which case ...."

File: pyaedt/generic/design_types.py [hunk 4]
```diff
@@ -1750,45 +1750,63 @@ def Siwave(
 }
 
 
-def get_pyaedt_app(project_name=None, design_name=None):
+def get_pyaedt_app(project_name=None, design_name=None, desktop=None):
     """Returns the Pyaedt Object of specific project_name and design_name.
```
Comment (PipKat) [inline pyaedt/generic/design_types.py:1754]:
```suggestion
    """Gets the PyAEDT object with a given project name and design name.
```

File: pyaedt/generic/design_types.py [hunk 5]
```diff
@@ -1750,45 +1750,63 @@ def Siwave(
...
+        oProject = odesktop.SetActiveProject(project_name)
+    if not oProject:
+        raise AttributeError("No Project Present.")
+    design_names = []
+    deslist = list(oProject.GetTopDesignList())
+    for el in deslist:
+        m = re.search(r"[^;]+$", el)
+        design_names.append(m.group(0))
+    if design_name and design_name not in design_names:
+        raise AttributeError("Design  {} doesn't exists in current Project.".format(design_name))
```
Comment (PipKat) [inline pyaedt/generic/design_types.py:1800]:
```suggestion
        raise AttributeError("Design  {} doesn't exist in current Project.".format(design_name))
```

File: pyaedt/generic/design_types.py [hunk 6]
```diff
@@ -1750,45 +1750,63 @@ def Siwave(
...
+    else:
+        raise AttributeError("No Desktop Present.")
+    if project_name and project_name not in odesktop.GetProjectList():
+        raise AttributeError("Project  {} doesn't exist in current Desktop.".format(project_name))
+    if not project_name:
+        oProject = odesktop.GetActiveProject()
+    else:
+        oProject = odesktop.SetActiveProject(project_name)
+    if not oProject:
+        raise AttributeError("No Project Present.")
```
Comment (PipKat) [inline pyaedt/generic/design_types.py:1793]:
```suggestion
        raise AttributeError("No project is present.")
```

File: pyaedt/generic/design_types.py [hunk 7]
```diff
@@ -1750,45 +1750,63 @@ def Siwave(
...
+                odesktop = desktop.odesktop
+                break
+    elif _desktop_sessions:
+        odesktop = list(_desktop_sessions.values())[-1]
+    elif "oDesktop" in dir(sys.modules["__main__"]):  # ironpython
+        odesktop = sys.modules["__main__"].oDesktop  # ironpython
+    else:
+        raise AttributeError("No Desktop Present.")
+    if project_name and project_name not in odesktop.GetProjectList():
+        raise AttributeError("Project  {} doesn't exist in current Desktop.".format(project_name))
```
Comment (PipKat) [inline pyaedt/generic/design_types.py:1787]:
```suggestion
        raise AttributeError("Project  {} doesn't exist in current desktop.".format(project_name))
```

File: pyaedt/generic/design_types.py [hunk 8]
```diff
@@ -1750,45 +1750,63 @@ def Siwave(
...
+        m = re.search(r"[^;]+$", el)
+        design_names.append(m.group(0))
+    if design_name and design_name not in design_names:
+        raise AttributeError("Design  {} doesn't exists in current Project.".format(design_name))
+    if not design_name:
+        oDesign = oProject.GetActiveDesign()
+    else:
+        oDesign = oProject.SetActiveDesign(design_name)
+    if not oDesign:
+        raise AttributeError("No Design Present.")
```
Comment (PipKat) [inline pyaedt/generic/design_types.py:1806]:
```suggestion
        raise AttributeError("No design is present.")
```

File: pyaedt/generic/general_methods.py [hunk 9]
```diff
@@ -210,10 +213,12 @@ def wrapper(*args, **kwargs):
             except MethodNotSupportedError:
                 message = "This Method is not supported in current AEDT Design Type."
```
Comment (PipKat) [inline pyaedt/generic/general_methods.py:214]:
```suggestion
                message = "This method is not supported in current AEDT design type."
```

File: pyaedt/generic/general_methods.py [hunk 10]
```diff
@@ -210,10 +213,12 @@ def wrapper(*args, **kwargs):
...
                 if settings.enable_screen_logs:
-                    print("**************************************************************")
-                    print("pyaedt error on Method {}:  {}. Please Check again".format(user_function.__name__, message))
-                    print("**************************************************************")
-                    print("")
+                    pyaedt_logger.logger.error("**************************************************************")
+                    pyaedt_logger.logger.error(
+                        "pyaedt error on Method {}:  {}. Please Check again".format(user_function.__name__, message)
```
Comment (PipKat) [inline pyaedt/generic/general_methods.py:218]:
```suggestion
                        "PyAEDT error on method {}:  {}. Check again".format(user_function.__name__, message)
```

File: pyaedt/generic/settings.py [hunk 11]
```diff
@@ -217,6 +219,28 @@ def aedt_version(self, value):
...
+    @aedt_install_dir.setter
+    def aedt_install_dir(self, value):
+        self._aedt_install_dir = value
+
+    @property
+    def use_multi_desktop(self):
+        """Enable multiple desktop sessions in same python script.
+        Current limitations are: release without closing desktop is not possible,
+        first desktop created has to be the last to be closed.
+        This is a beta feature."""
```
Comment (PipKat) [inline pyaedt/generic/settings.py:236]:
```suggestion
        """Flag indicating if multiple desktop sessions are enabled in the same Python script.
        Current limitations follow:
        
         - Release without closing the desktop is not possible,
        - The first desktop created must be the last to be closed.

        Enabling multiple desktop sessions is a beta feature."""
```

**Comment by Samuelopez-ansys** [comment]
@maxcapodi78 Do you need the global variable python_grpc_wrapper?

**Review by Samuelopez-ansys — APPROVED** [review]

**Pull request merged**